### ISSUE TITLE
Files dropped after SyncPreCheckpoint but before checkpoint record is written might be left orphaned

### ISSUE THREAD

**Issue by huansong** [greenplum-db/gpdb#16120]
## Bug Report
After `SyncPreCheckpoint` but before the checkpoint record is written to WAL, there's a small window in which if one tries to drop a main relfilenode, the unlink request can be forgotten if the server crash/restart w/o another checkpoint. The DROP won't be replayed during server recovery, hence the file will be left orphaned.

This case is pretty rare and I guess not entirely unexpected because orphaned file due to server crash is a well known issue. But this particular scenario seems new. It was found in CI where `gpstop -air` restarts the server w/o a checkpoint. For now we'll avoid the `-i` option in the corresponding restarting command to mitigate the flakiness. 

### Greenplum version or build
7X (HEAD 141c719b6bf374a229dbe53e7296182e6b76d26c). Not GPDB-specific - upstream has it too.

### Step to reproduce the behavior

1. Firstly, add this fault in code:
```
--- a/src/backend/access/transam/xlog.c
+++ b/src/backend/access/transam/xlog.c
@@ -9202,6 +9202,7 @@ CreateCheckPoint(int flags)
         */
        SyncPreCheckpoint();

+       SIMPLE_FAULT_INJECTOR("after_sync_pre_checkpoint");
        /*
         * Use a critical section to force system panic if we have trouble.
         */
```
2. Run this over a connection and let it be suspended.
```sql
create table t1(a int);
select gp_inject_fault('after_sync_pre_checkpoint', 'suspend', 2);
checkpoint;
```

3. Run this over another connection.
```sql
drop table t1;
select gp_inject_fault('after_sync_pre_checkpoint', 'reset', 2);
```

4. Restart the server using immediate option (-i)
```sql
gpstop -ari
```

5. Check orphaned file, should see one.
```sql
select * from gp_toolkit.gp_check_orphaned_files;
```

**Comment by interma** [comment]
> hence the file will be left orphaned.

One question:
Will the orphaned files be unlinked later (by some operations)? or keep it forever in the data directory? (I took a look at 'gp_check_files' case before: it reported some but I didn't find them in the whole disk).
Thanks, huansong.

**Comment by huansong** [comment]
> I took a look at 'gp_check_files' case before: it reported some but I didn't find them in the whole disk

Yeah I noticed that too, and in some other CI instance the orphaned files were still there. It was really confusing. But I found that the files were only missing in CI instances that executed more tests than `regress`. In other CI instances where only `regress` tests were run, the files were alway there. 

I think what probably happened is that in some tests we intentionally fail a primary and promote the mirror, then do gprecoverseg to bring the old primary back and make it primary again (we do that a lot in `isolation2`). During this process, the old primary will be sync'ed (pg_rewind) with the old mirror, and since the mirror doesn't have the orphaned file (the mirror just replays WALs so it has a consistent view of files), the orphaned files on the old primary would be removed.

**Comment by interma** [comment]
Thank you for the explanation.
It helps a lot!
